### DISCUSSION TITLE
Welcome to geodesyNets Discussions!

### DISCUSSION THREAD

**Discussion by darioizzo** [darioizzo/geodesyNets#108] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by LWTongf** [comment]
Dear author, I have a few questions. In the following document, I want to ask you for help.
[question.docx](https://github.com/darioizzo/geodesyNets/files/10840799/question.docx)

**Reply by darioizzo** [reply]
Dear @LWTongf thanks for your interest in the project. In order to be able to easily answer you please format your question directly in the discussion forum, avoid using an intermediate document. Make sure to properly format your code examples making use of the  markup syntax.

**Comment by LWTongf** [comment]
Dear author:
shalom!
I am a student. After reading your code, I want to ask you some questions. The code is as follows:
def ACC_trap (target_points , model , encoding=direct_encoding(), N=10000, verbose=False , noise=1e-5, sample_points=None, h=None, domain=None ):
"""Uses a 3D trapezoid rule for the evaluation of the integral in the potential from the modeled density
Args:
target_points (2-D array-like): a (N,3) array-like object containing the points.
model (callable (a,b)->1): neural model for the asteroid.
encoding: the encoding for the neural inputs.
 N (int): number of points.If a grid is passed should match that
verbose (bool, optional): Print intermediate results.Defaults to False.
noise (float): random noise added to point positions.
sample_points (torch tensor): grid to sample the integral on
h (float): grid spacing, only has to be passed if grid is passed.
domain (torch.tensor): integration domain [3,2] , pass None for [-1,1]^3
Returns:
Tensor: Computed potentials per point
"""
if domain is None :  # None might be passed as well
domain = [[-1, 1], [-1, 1], [-1, 1]]
#init result vector
retval = torch.empty(len (target_points), 3,
device =os.environ["TORCH_DEVICE"])
#Determine grid to compute on
if sample_points is None :
sample_points , h , N = compute_integration_grid(N , noise , domain)
else :
if h is None :
raise (ValueError ("h has to be passed if sample points are passed."))
#Evaluate Rho on the grid
rho = _compute_model_output(model , encoding , sample_points)
for i , target_point in enumerate (target_points):
#Evaluate all points
distance = torch.sub(target_point , sample_points)
f_values = (rho /
torch.pow(torch.norm(distance , dim =1), 3).view(-1, 1) * distance)
evaluations = f_values.reshape([N , N , N , 3])  # map to z,y,x
#area = h / 2 * (f0 + f2)
int_x = h[0] / 2 * (evaluations[:, :, 0:-1, :] +
evaluations[:, :, 1:, :])
int_x = torch.sum(int_x , dim =2)
int_y = h[1] / 2 * (int_x[:, 0:-1, :] + int_x[:, 1:, :])
int_y = torch.sum(int_y , dim =1)
int_z = h[2] / 2 * (int_y[0:-1, :] + int_y[1:, :])
int_z = torch.sum(int_z , dim =0)
retval[i] = int_z
return -retval

Now I would like to ask you to intercept the code section:
for i , target_point in enumerate (target_points):
#Evaluate all points
distance = torch.sub(target_point , sample_points)
f_values  = (rho /
torch.pow(torch.norm(distance , dim =1), 3).view(-1, 1) * distance)
evaluations = f_values.reshape([N , N , N , 3])  # map to z,y,x
#area = h / 2 * (f0 + f2)
int_x = h[0] / 2 * (evaluations[:, :, 0:-1, :] +
evaluations[:, :, 1:, :])
int_x = torch.sum(int_x , dim =2)
int_y = h[1] / 2 * (int_x[:, 0:-1, :] + int_x[:, 1:, :])
int_y = torch.sum(int_y , dim =1)
int_z = h[2] / 2 * (int_y[0:-1, :] + int_y[1:, :])
int_z = torch.sum(int_z , dim =0)
retval[i] = int_z
return -retval
In the paper, you use the following formula when calculating the predicted value:
a(r)= G\int_{x\epsilon V}^{}\frac{\rho (x)}{|r-x|^{3}}(**r**-**x**)dv
In your code, you use the 3 D trapezoidal integration rule, and you first calculate the distance increment between the target point and the sampling point:
distance = torch.sub(target_point , sample_points)
You found the r based on this:
torch.pow(torch.norm(distance , dim =1), 3).view(-1, 1)
Then the density you predicted is worth the f_values in the code:
f_values = (rho  /torch.pow(torch.norm(distance , dim =1), 3).view(-1, 1) * distance)
Here rho is the value of the sampling point input encoded into the model. The code is as follows:
def _compute_model_output (model , encoding , sample_points):
"""Computes model output on the passed points using the passed encoding
Args:
model (torch model): neural network to eval
encoding (encoding): encoding for network input (dim has to match)
sample_points (torch tensor): points to sample at
Returns:
torch tensor: computed values
"""
#check dimensions match
_check_model_encoding_compatibility(model , encoding)
#1 - compute the inputs to the ANN encoding the sampled points
nn_inputs = encoding(sample_points)
#2 - check if any values were NaN
if torch.any(nn_inputs != nn_inputs):
warnings.warn("The network generated NaN outputs!")
nn_inputs[nn_inputs != nn_inputs] = 0.0  # set Nans to 0
#3 - compute the predicted density at the points
return model(nn_inputs)
# I would like to ask you in the above reasoning is correct situation, I will you this code:
f_values  = (rho /torch.pow(torch.norm(distance , dim =1), 3).view(-1, 1) * distance)
Equivalence is to the following formula:
f_values=\frac{rho}{r^{3}}**r**
among:
r=distance^{2}
distance=target_point-sample_point
# Why did the x in the formula given in your paper disappear here (question 1)?(The part I mark in red):
\frac{\rho (x)}{|r-{\color{Red} x}|^{3}}(**r**-{\color{Red} **x**})
# Next you will map to z, y, x. I ask the purpose of this (question 2)?(Code below)
evaluations = f_values.reshape([N , N , N , 3])  # map to z,y,x
# Why not map to x, y, and z (question 3)?
Then you used the 3 D trapezoidal integration rule, I don't understand your integral side (code:)
f_values = (rho /
torch.pow(torch.norm(distance , dim =1), 3).view(-1, 1) * distance)
evaluations = f_values.reshape([N , N , N , 3])  # map to z,y,x
#area = h / 2 * (f0 + f2)
int_x = h[0] / 2 * (evaluations[:, :, 0:-1, :] +
evaluations[:, :, 1:, :])
int_x = torch.sum(int_x , dim =2)
int_y = h[1] / 2 * (int_x[:, 0:-1, :] + int_x[:, 1:, :])
int_y = torch.sum(int_y , dim =1)
int_z = h[2] / 2 * (int_y[0:-1, :] + int_y[1:, :])
int_z = torch.sum(int_z , dim =0)
retval[i] = int_z
# Can I understand here that you first integrate it over N small cubes? Or can you send me the integral formula or steps you use as a manuscript about the idea of your integral (question 4)?
In your following module (code below):
def U_mc (target_points , model , encoding=direct_encoding(), N=3000, domain=None ):
"""Plain Monte Carlo evaluation of the potential from the modelled density

Args:
target_points (2-D array-like): a (N,3) array-like object containing the points.
model (callable (a,b)->1): neural model for the asteroid.
encoding: the encoding for the neural inputs.
 N (int): number of points.
domain (torch.tensor): integration domain [3,2] , pass None for [-1,1]^3. Currently Not Implemented!
"""
if domain is not None :
raise NotImplementedError(
"Custom domain is not yet implemented for U_mc.")
#init result vector
retval = torch.empty(len(target_points), 1)
#We generate randomly points in the [-1,1]^3 bounds
sample_points = torch.rand(N , 3, device =os.environ["TORCH_DEVICE"]) * 2 - 1
#Evaluate Rho on the points
rho = _compute_model_output(model , encoding , sample_points)
#Only for the points inside we accumulate the integrand (MC method)
#For high-dimensional volume integration formula is: F= (v / N) * sum (f (x))
for i , target_point in enumerate(target_points):
retval[i] = torch.sum(
rho/torch.norm(target_point - sample_points , dim =1).view(-1, 1)) / N
return  - 8 * retval
# You used the Monte Carlo integral method, and the formula I found is:

f=\frac{V}{N}\sum_{i=0}^{N}f(x_{i})

This corresponds to your code in the module as follows:
retval[i] = torch.sum(rho/torch.norm(target_point - sample_points , dim =1).view(-1, 1)) / N
# Is this true (Question 5)? Did you omit V for a volume of 1 (Question 6)? I would like you to send me your integral formula for Monte Carlo.And why did you end up returning a-8 * retval (Question 7)?
In this module:
def _sample_cubical (N , scale_bounds=[1.1, 1.2]):
"""Generates N uniform random samples from a cube with passed scale.All points outside unit cube.
 Args:
N (int): Nr of points to create.
scale_bounds (float, optional): Scales of the domain for the points.Defaults to [1.1, 1.2].
 Returns:
Torch tensor: Sampled points
"""
#Approximation of percentage points in Unitsphere
approx = (scale_bounds[0] / scale_bounds[1])**3
#Sample twice the expected number of points necessary to achieve N
approx_necessary_samples = int (2 * N * (1.0 / (1.0 - approx)))
points = (torch.rand(approx_necessary_samples , 3,
device =os.environ["TORCH_DEVICE"])*2 - 1)*scale_bounds[1]
#Discard points inside unitcube
domain = scale_bounds[0]*torch.tensor(
[[-1, 1], [-1, 1], [-1, 1]], device =os.environ["TORCH_DEVICE"])
points = _limit_to_domain(points , domain)
#Take first N points (N.B.that in super unlikely event of
#less than N points this will not crash.I tried.:))
return points[:N]
I don't quite understand your sampling method, so I want to ask you about this code:
approx_necessary_samples = int (2 * N * (1.0 / (1.0 - approx)))
points = (torch.rand(approx_necessary_samples , 3
device =os.environ["TORCH_DEVICE"])*2 - 1)*scale_bounds[1]
# Why double the number of points expected (Question 8)? And what is the principle of obtaining the sampling points in the way you define it (question 9)? What I understand is that you used a linear mapping relationship to easily reproduce the results, is that correct?
In this module:
def _get_altitude_sampler (N , altitude , limit_shape_to_asteroid , plot_normals=False, discard_points_inside=True, replace=True ):
"""This creates a sampler that samples from the triangle centers of the passed mesh + their normal
Args:
N (int): number of points to get per batch
altitude (float): altitude above the faces
limit_shape_to_asteroid (str): path of to asteroid mesh
plot_normals (bool, optional): Display normals and created points for debugging.Defaults to False.
discard_points_inside (bool, optional): Will discard all points that lie inside the asteroid (can happen for nonconvex ones).Defaults to True.
replace (bool, optional): If points are allowed to be sampled twice in the same batch or not
(for false maximum sample points = #triangles in mesh)
Returns:
func: sampler function
"""
#print("Sampling ", limit_shape_to_asteroid, " at altitude=", altitude)
#Load asteroid triangles
with open (limit_shape_to_asteroid , "rb") as file:
mesh_vertices , mesh_triangles = pk.load(file)
mesh_faces = [[3, t[0], t[1], t[2]] for t in mesh_triangles]
#Create PV Polydata
mesh = pv.PolyData(np.asarray(mesh_vertices), np.asarray(mesh_faces))
#mesh.plot(show_scalar_bar=False)
mesh.compute_normals(cell_normals =True, point_normals =False,
inplace =True, consistent_normals =False,
non_manifold_traversal =False )
centers = mesh.cell_centers().points
points_at_altitude = centers + altitude * mesh['Normals']
if plot_normals:
#mesh.plot()
pv.set_plot_theme("document")
plotter = pv.Plotter()
point_cloud = pv.PolyData(centers)
point_cloud["vectors"] = mesh['Normals']
arrows = point_cloud.glyph(
orient ='vectors', scale =True, factor =0.1)
plotter.add_mesh(mesh , color =[0.2, 0.2, 0.2], lighting =True,
smooth_shading =False, show_edges =True, edge_color =[0.25, 0.25, 0.25])
plotter.add_mesh(arrows , opacity =1.0, color ='lightblue')
plotter.add_mesh(points_at_altitude , style ="points",
color ='red', point_size =4, render_points_as_spheres =True )
plotter.show_grid()
#plotter.show()
plotter.show(screenshot ="figures/sampler.png", window_size =[1200, 800])
#Discard points that are too close (or too far)
eps = 1e-4  # maximum altitude error
kd_tree = KDTree(centers)
distances , _ = kd_tree.query(points_at_altitude , k =1)
distance_correct = np.abs(altitude-distances) < eps
print ("Discarding " + str (len (distance_correct) - np.sum(distance_correct)) + " of " + str (len (distance_correct)) +
" points in altitude sampler which did not meet requested altitude.")
points_at_altitude = points_at_altitude[distance_correct]
if discard_points_inside:
#print("Discarding points inside asteroid.Prev len=",
#len(points_at_altitude))
triangles = unpack_triangle_mesh(mesh_vertices , mesh_triangles)
points_at_altitude = points_at_altitude[is_outside_torch(
torch.tensor(points_at_altitude).float().detach(), triangles).detach().cpu().numpy()]
del triangles
#print("Now len=", len(points_at_altitude))
return lambda : torch.tensor(points_at_altitude[np.random.choice(
points_at_altitude.shape[0], N , replace =replace), :])
I don't quite understand the following code:
mesh_faces = [[3, t[0], t[1], t[2]] for t in mesh_triangles]
# Can you explain what it means (Question 10)? Where does the 3 come from in this case (question 11)?
Above is my confusion, I would like to ask you to give me guidance, and finally I wish you a smooth work, a happy life.
